### ISSUE TITLE
[Request] JSON Export Format

### ISSUE THREAD

**Issue by sadsfae** [Alumian/CEPGP-Retail#127]
I'm new to CEPGP and was trying to import logs from the addon to EQDKP Plus and get the following error from EQDKP Plus:

https://download.eqdkp-plus.eu/Extensionlist/Epgp-import-p586/?

```
EPGP-Log could not be imported, maybe it is damaged, has a wrong format or a newer one has already been imported
```

I am only exporting this via the `export standings` button based on guild roster and existing office notes which contain values

I realize that this is for CEPGP and not 3rd party log importers but I wanted to confirm a few things:

1) What is the file structure of the CSV export?
2) Is there another CSV field for event/encounter I could try?

My file format looks like this, and was created by the addon:

```
Sneakums,Warlock,Member,0,1,0
Spellcaster,Mage,Initiate,0,1,0
Sprunky,Rogue,Member,16,29,0.55
```

Has anyone else gotten this working?  Just trying to get an initial import working.

**Comment by Alumian** [comment]
If the import isn't working then that would suggest that EQDKP is expecting a structure other than CSV. CSV literally means comma separated values, so the data structure as you can see in your example is name, class, rank, EP, GP, PR.

**Comment by sadsfae** [comment]
According to one of the EQDKP devs the old EPGP plugin it was based on used JSON strings:

https://forum.eqdkp-plus.eu/thread/4893-help-with-wow-classic-and-epgp-log-importer-extension/

I am looking for some example logs from this older program, perhaps it's possible to convert the CSV output into JSON strings?

**Comment by Alumian** [comment]
I'm sure you could but I'm happy to just write in a new data structure to the export function so you don't have to do the conversion. Are you able to supply me with a scaffold of how you need the data to look? For example, like I said for CSV, the scaffold was name, class, rank, EP, GP, PR.

**Comment by sadsfae** [comment]
> I'm sure you could but I'm happy to just write in a new data structure to the export function so you don't have to do the conversion. Are you able to supply me with a scaffold of how you need the data to look? For example, like I said for CSV, the scaffold was name, class, rank, EP, GP, PR.

That would be fantastic, I am digging around looking for an example of the old EPGP JSON export format it seems to want now.

**Comment by GodMod** [comment]
The format looked like this:
```
{"guild":"Circle of Nine","region":"eu","min_ep":3000,"base_gp":1000,"roster":[["Dekann",990,1000],["Inspyre",2950,1000],["Sinajnsoul",1968,1000],["Stegs",24174,3520],["Badylq",18908,1759]],"timestamp":1353152700,"extras_p":0,"realm":"Burning Legion"}
```
The first value after the name was the EP, the second the GP.  Important for the import have only bin the roster value, as well as the timestamp value to prevent the import of same standings.

**Comment by Alumian** [comment]
Okay so are all of those fields required? i.e. guild, region, min ep, base gp?

**Comment by GodMod** [comment]
No, for the import required are roster and timestamp. Thats what i meant with important, sry.

**Comment by sadsfae** [comment]
I'm standing by to test JSON strings with the existing plugin if it helps, or I can give you access to our EQDKP site for testing purposes.  I'm sure a lot of other folks will find use for this.

**Comment by Alumian** [comment]
Okay I've just written the new format function. Are you on the CEPGP discord? Can you drop me a message? I'll send you a copy for testing.

**Comment by sadsfae** [comment]
> Okay I've just written the new format function. Are you on the CEPGP discord? Can you drop me a message? I'll send you a copy for testing.

I am on the discord now, @repo#3449 happy to upgrade and test.

**Comment by sadsfae** [comment]
[Here is the JSON dump](https://gist.github.com/sadsfae/5d7ff7d7bcc6a21f9860a5784aaeea08) from the testing CEPGP addon that's built.

I am still getting the same error on the EQDKP side but the data looks right to my eyes.

```
EPGP-Log could not be imported, maybe it is damaged, has a wrong format or a newer one has already been imported.
```

@GodMod @Alumian I will enable debug logging to see if I can dig up anything that way, in the meantime is there anything missing?

**Comment by sadsfae** [comment]
**Update:**

I've enabled `debugging: 4` in settings as well as in my `php.ini` but see nothing wrong on the server side for the log import.

Looking through the plugin code it seems to hit this:

https://github.com/EQdkpPlus/plugin-epgpimport/blob/master/admin/import.php#L54

`$mixedResult` is defined here if I'm reading it correctly.

https://github.com/EQdkpPlus/plugin-epgpimport/blob/master/admin/import.php#L54

**Comment by Alumian** [comment]
@GodMod can you see any issues with the export string that sadsfae has supplied?

**Comment by GodMod** [comment]
The Output is invalid JSON. There is a `,` to much after the last member. Just check it with a tool like https://jsonformatter.curiousconcept.com/

**Comment by sadsfae** [comment]
> The Output is invalid JSON. There is a `,` to much after the last member. Just check it with a tool like https://jsonformatter.curiousconcept.com/

@GodMod @Alumian can confirm this works stripping off the last `,` from the last member and removing all of the line breaks from the JSON output.

[Here is the modified JSON payload](https://gist.github.com/sadsfae/160564fa558e8ce2772248b964858db2)

Here is what it looks like after import, note that this is a clean EQDKP installation.

https://wow.deademperors.com/index.php/Points/?

@Alumian can we try also pulling in class and guild rank too where appropriate?  @GodMod are these accepted values?  It'd be nice as it can easily populate your guild roster based on the CEPGP in-game "Export Standings" functionality with the WoW Addon-on.

**Comment by Fjarnskaggl** [comment]
`function CEPGP_formatExport(form)
	--form is the export format
	if form == "CSV" then
		local temp = {};
		local text = "";
		local size = CEPGP_ntgetn(CEPGP_roster);
		for k, v in pairs(CEPGP_roster) do
			temp[CEPGP_ntgetn(temp)+1] = {
				[1] = k,
				[2] = v[2],
				[3] = v[3],
				[4] = v[4],
				[5] = v[5],
				[6] = v[6]
			};
		end
		temp = CEPGP_tSort(temp, 1);
		for i = 1, size do
			text = text .. temp[i][1] .. "," .. temp[i][2] .. "," .. temp[i][3] .. "," .. temp[i][5] .. "," .. temp[i][6] .. "\n"; --Line 16
		end
		_G["CEPGP_export_dump"]:SetText(text);
		_G["CEPGP_export_dump"]:HighlightText();
		temp = nil;
		text = nil;
		size = nil;
	elseif form == "JSON" then
		local temp = {};
		local text = "";
		text = "{\n";
		text = text .. "\"roster\": [\n";
		for name, values in pairs(CEPGP_roster) do
			local EP, GP = CEPGP_getEPGP(values[5]);
			temp[#temp+1] = {
				[1] = name,
				[2] = EP,
				[3] = GP
			};
		end
		temp = CEPGP_tSort(temp, 1);
		tlen = CEPGP_ntgetn(temp)
		for i = 1, #temp do
			if  i ~= tlen then
			text = text .. "[\"" .. temp[i][1] .. "\"," .. temp[i][2] .. "," .. temp[i][3] .. "],\n";
			else
				text = text .. "[\"" .. temp[i][1] .. "\"," .. temp[i][2] .. "," .. temp[i][3] .. "]\n";
			end
		end
		text = text .. "],\n";
		text = text .. "\"timestamp\":" .. time() .. "\n";
		text = text .. "}";
		_G["CEPGP_export_dump"]:SetText(text);
		_G["CEPGP_export_dump"]:HighlightText();
	end
end`
this resolves the issue with the `,` being added to the end of the list

**Comment by Fjarnskaggl** [comment]
```LUA
function CEPGP_formatExport(form)
	--form is the export format
	if form == "CSV" then
		local temp = {};
		local text = "";
		local size = CEPGP_ntgetn(CEPGP_roster);
		for k, v in pairs(CEPGP_roster) do
			temp[CEPGP_ntgetn(temp)+1] = {
				[1] = k,
				[2] = v[2],
				[3] = v[3],
				[4] = v[4],
				[5] = v[5],
				[6] = v[6]
			};
		end
		temp = CEPGP_tSort(temp, 1);
		for i = 1, size do
			text = text .. temp[i][1] .. "," .. temp[i][2] .. "," .. temp[i][3] .. "," .. temp[i][5] .. "," .. temp[i][6] .. "\n"; --Line 16
		end
		_G["CEPGP_export_dump"]:SetText(text);
		_G["CEPGP_export_dump"]:HighlightText();
		temp = nil;
		text = nil;
		size = nil;
	elseif form == "JSON" then
		local temp = {};
		local text = "";
		text = "{\n";
		text = text .. "\"roster\": [\n";
		for name, values in pairs(CEPGP_roster) do
			local EP, GP = CEPGP_getEPGP(values[5]);
			temp[#temp+1] = {
				[1] = name,
				[2] = EP,
				[3] = GP
			};
		end
		temp = CEPGP_tSort(temp, 1);
		tlen = CEPGP_ntgetn(temp)
		for i = 1, #temp do
			if  i ~= tlen then
			text = text .. "[\"" .. temp[i][1] .. "\"," .. temp[i][2] .. "," .. temp[i][3] .. "],\n";
			else
				text = text .. "[\"" .. temp[i][1] .. "\"," .. temp[i][2] .. "," .. temp[i][3] .. "]\n";
			end
		end
		text = text .. "],\n";
		text = text .. "\"timestamp\":" .. time() .. "\n";
		text = text .. "}";
		_G["CEPGP_export_dump"]:SetText(text);
		_G["CEPGP_export_dump"]:HighlightText();
	end
end```

Resolves the issue with the extra `,`

**Comment by Alumian** [comment]
Yep Ive already written a fix for it. I'll push a new beta in the coming days when I've added more fixes and changes.

**Comment by sadsfae** [comment]
> ```lua
> function CEPGP_formatExport(form)
> 	--form is the export format
> 	if form == "CSV" then
> 		local temp = {};
> 		local text = "";
> 		local size = CEPGP_ntgetn(CEPGP_roster);
> 		for k, v in pairs(CEPGP_roster) do
> 			temp[CEPGP_ntgetn(temp)+1] = {
> 				[1] = k,
> 				[2] = v[2],
> 				[3] = v[3],
> 				[4] = v[4],
> 				[5] = v[5],
> 				[6] = v[6]
> 			};
> 		end
> 		temp = CEPGP_tSort(temp, 1);
> 		for i = 1, size do
> 			text = text .. temp[i][1] .. "," .. temp[i][2] .. "," .. temp[i][3] .. "," .. temp[i][5] .. "," .. temp[i][6] .. "\n"; --Line 16
> 		end
> 		_G["CEPGP_export_dump"]:SetText(text);
> 		_G["CEPGP_export_dump"]:HighlightText();
> 		temp = nil;
> 		text = nil;
> 		size = nil;
> 	elseif form == "JSON" then
> 		local temp = {};
> 		local text = "";
> 		text = "{\n";
> 		text = text .. "\"roster\": [\n";
> 		for name, values in pairs(CEPGP_roster) do
> 			local EP, GP = CEPGP_getEPGP(values[5]);
> 			temp[#temp+1] = {
> 				[1] = name,
> 				[2] = EP,
> 				[3] = GP
> 			};
> 		end
> 		temp = CEPGP_tSort(temp, 1);
> 		tlen = CEPGP_ntgetn(temp)
> 		for i = 1, #temp do
> 			if  i ~= tlen then
> 			text = text .. "[\"" .. temp[i][1] .. "\"," .. temp[i][2] .. "," .. temp[i][3] .. "],\n";
> 			else
> 				text = text .. "[\"" .. temp[i][1] .. "\"," .. temp[i][2] .. "," .. temp[i][3] .. "]\n";
> 			end
> 		end
> 		text = text .. "],\n";
> 		text = text .. "\"timestamp\":" .. time() .. "\n";
> 		text = text .. "}";
> 		_G["CEPGP_export_dump"]:SetText(text);
> 		_G["CEPGP_export_dump"]:HighlightText();
> 	end
> end
> ```
> 
> Resolves the issue with the extra `,`

Thanks for that.  @Alumian has a test build that corrects it now that I can confirm works.

Also, EQDKP released version `0.3.0` of the `EPGP-Import` plugin which now works with the existing CSV format (even with line wrapping) today.

https://download.eqdkp-plus.eu/Extensionlist/Epgp-import-p586/?

**Comment by Alumian** [comment]
Implemented in https://github.com/Alumian/CEPGP-Retail/releases/tag/1.12.7
